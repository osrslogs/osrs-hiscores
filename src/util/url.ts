import { Mode, SkillName, ActivityName } from '../types';
import { SKILLS, ACTIVITIES } from './constants';

const urlBase = 'https://secure.runescape.com';
const urlStats = 'index_lite.ws';
const urlScores = 'overall.ws';
const urlMode: { [key: string]: string } = {
  normal: 'm=hiscore_oldschool',
  ironman: 'm=hiscore_oldschool_ironman',
  ultimate: 'm=hiscore_oldschool_ultimate',
  hardcore: 'm=hiscore_oldschool_hardcore_ironman',
  deadman: 'm=hiscore_oldschool_deadman',
  seasonal: 'm=hiscore_oldschool_seasonal',
  tournament: 'm=hiscore_oldschool_tournament',
};

/**
 * Returns the correct url mode
 *
 * @param {Mode} [mode] The game mode to select. Defaults to `normal`
 *
 * @returns {string} The urlMode property
 */
const getUrlModeWithDefault = (mode: Mode = 'normal'): string => urlMode[mode];

/**
 * Builds the url string for personal stats
 *
 * @param {string} player The player name to lookup
 * @param {Mode} [mode] The game mode to select. Defaults to `normal`
 *
 * @returns {string}
 */
export const buildStatsUrl = (player: string, mode?: Mode): string => {
  const urlStatsPlayer = urlStats.concat(`?player=${player}`);
  return [urlBase, getUrlModeWithDefault(mode), urlStatsPlayer].join('/');
};

/**
 * Builds the url string for skill page
 *
 * @param {SkillName} skill The skill name to lookup
 * @param {Mode} [mode] The mode to select. Defaults to `normal`
 *
 * @returns {string}
 */
export const buildSkillPageUrl = (skill: SkillName, mode?: Mode): string => {
  const tableIdx = SKILLS.indexOf(skill);
  const urlScoresTable = urlScores.concat(`?table=${tableIdx}`);
  return [urlBase, getUrlModeWithDefault(mode), urlScoresTable].join('/');
};

/**
 * Builds the url string for activity page
 *
 * @param {ActivityName} activity The activity name to lookup
 * @param {Mode} [mode] The mode to select. Defaults to `normal`
 *
 * @returns {string}
 */
export const buildActivityPageUrl = (activity: ActivityName, mode?: Mode): string => {
  const tableIdx = ACTIVITIES.indexOf(activity);
  const urlScoresTable = urlScores.concat(`?category_type=1&table=${tableIdx}`);
  return [urlBase, getUrlModeWithDefault(mode), urlScoresTable].join('/');
};

/**
 * Builds the url string for formatted player name (highlighted in table)
 *
 * @param {string} player The player to lookup
 * @param {SkillName} [skill] The skill table to use. Defaults to `overall`
 * @param {Mode} [mode] The mode to select. Defaults to `normal`
 *
 * @returns {string}
 */
export const buildSkillPageWithDisplayNameUrl = (player: string, skill?: SkillName, mode?: Mode): string => {
  const defaultOrSkill = skill || 'overall';
  const pageUrl = buildSkillPageUrl(defaultOrSkill, mode);
  return pageUrl.concat(`&user=${player}`);
};
