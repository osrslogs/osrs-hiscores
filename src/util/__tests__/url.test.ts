import { buildStatsUrl, buildSkillPageUrl, buildActivityPageUrl, buildSkillPageWithDisplayNameUrl } from '../url';

describe('buildStatsUrl', () => {
  it('returns expected url without mode specified', () => {
    const url = buildStatsUrl('zezima');
    expect(url).toMatchSnapshot();
  });

  it('returns expected url when mode is specified', () => {
    const url = buildStatsUrl('zezima', 'hardcore');
    expect(url).toMatchSnapshot();
  });
});

describe('buildSkillPageUrl', () => {
  it('returns expected url with default mode', () => {
    const url = buildSkillPageUrl('woodcutting');
    expect(url).toMatchSnapshot();
  });

  it('returns expected url with undefined mode', () => {
    const url = buildSkillPageUrl('woodcutting', undefined);
    expect(url).toMatchSnapshot();
  });

  it('returns expected url with specified mode', () => {
    const url = buildSkillPageUrl('hitpoints', 'hardcore');
    expect(url).toMatchSnapshot();
  });
});

describe('buildActivityPageUrl', () => {
  it('returns expected url with default mode', () => {
    const url = buildActivityPageUrl('cerberus');
    expect(url).toMatchSnapshot();
  });

  it('returns expected url with default mode', () => {
    const url = buildActivityPageUrl('cerberus', undefined);
    expect(url).toMatchSnapshot();
  });

  it('returns expected url with specified mode', () => {
    const url = buildActivityPageUrl('easy', 'hardcore');
    expect(url).toMatchSnapshot();
  });
});

describe('buildDisplayNameUrl', () => {
  it('returns expected url with default skill and mode', () => {
    const url = buildSkillPageWithDisplayNameUrl('zezima');
    expect(url).toMatchSnapshot();
  });

  it('returns expected url with default mode', () => {
    const url = buildSkillPageWithDisplayNameUrl('zezima', 'construction');
    expect(url).toMatchSnapshot();
  });

  it('returns expected url with default skill', () => {
    const url = buildSkillPageWithDisplayNameUrl('zezima', undefined, 'ironman');
    expect(url).toMatchSnapshot();
  });
});
