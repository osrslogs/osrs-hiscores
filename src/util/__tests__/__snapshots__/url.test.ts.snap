// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildActivityPageUrl returns expected url with default mode 1`] = `"https://secure.runescape.com/m=hiscore_oldschool/overall.ws?category_type=1&table=16"`;

exports[`buildActivityPageUrl returns expected url with default mode 2`] = `"https://secure.runescape.com/m=hiscore_oldschool/overall.ws?category_type=1&table=16"`;

exports[`buildActivityPageUrl returns expected url with specified mode 1`] = `"https://secure.runescape.com/m=hiscore_oldschool_hardcore_ironman/overall.ws?category_type=1&table=5"`;

exports[`buildDisplayNameUrl returns expected url with default mode 1`] = `"https://secure.runescape.com/m=hiscore_oldschool/overall.ws?table=23&user=zezima"`;

exports[`buildDisplayNameUrl returns expected url with default skill 1`] = `"https://secure.runescape.com/m=hiscore_oldschool_ironman/overall.ws?table=0&user=zezima"`;

exports[`buildDisplayNameUrl returns expected url with default skill and mode 1`] = `"https://secure.runescape.com/m=hiscore_oldschool/overall.ws?table=0&user=zezima"`;

exports[`buildSkillPageUrl returns expected url with default mode 1`] = `"https://secure.runescape.com/m=hiscore_oldschool/overall.ws?table=9"`;

exports[`buildSkillPageUrl returns expected url with specified mode 1`] = `"https://secure.runescape.com/m=hiscore_oldschool_hardcore_ironman/overall.ws?table=4"`;

exports[`buildSkillPageUrl returns expected url with undefined mode 1`] = `"https://secure.runescape.com/m=hiscore_oldschool/overall.ws?table=9"`;

exports[`buildStatsUrl returns expected url when mode is specified 1`] = `"https://secure.runescape.com/m=hiscore_oldschool_hardcore_ironman/index_lite.ws?player=zezima"`;

exports[`buildStatsUrl returns expected url without mode specified 1`] = `"https://secure.runescape.com/m=hiscore_oldschool/index_lite.ws?player=zezima"`;
